runtime/src/main/java/org/corfudb/runtime/view/StreamsView.java

```diff
@@ -132,90 +132,98 @@ public long append(@Nonnull Object object, @Nullable TxResolutionInfo conflictIn
-                // If we're here, we succeeded, return the acquired token.
-                return tokenResponse.getSequence();
-            } catch (OverwriteException oe) {
-                // We were overwritten, get a new token and try again.
-                log.warn("append[{}]: Overwritten after {} retries, streams {}",
-                        tokenResponse.getSequence(), x,
-                        Arrays.stream(streamIDs).map(Utils::toReadableId).collect(Collectors.toSet()));
-
-                if (conflictInfo != null) {
-                    // On retry, check for conflicts only from the previous attempt position,
-                    // otherwise the transaction will always conflict with itself.
-                    conflictInfo.setSnapshotTimestamp(tokenResponse.getToken());
+                if (abortCause != null) {
+                    throw new TransactionAbortedException(
+                            conflictInfo,
+                            tokenResponse.getConflictKey(), tokenResponse.getConflictStream(),
+                            tokenResponse.getToken().getSequence(), abortCause,
+                            TransactionalContext.getCurrentContext());
                 }
 
-            } catch (StaleTokenException se) {
-                // the epoch changed from when we grabbed the token from sequencer
-                log.warn("append[{}]: StaleToken, streams {}", tokenResponse.getSequence(),
-                        Arrays.stream(streamIDs).map(Utils::toReadableId).collect(Collectors.toSet()));
-
-                throw new TransactionAbortedException(
-                        conflictInfo,
-                        tokenResponse.getConflictKey(), tokenResponse.getConflictStream(),
-                        tokenResponse.getToken().getSequence(),
-                        AbortCause.NEW_SEQUENCER, // in the future perhaps define a new AbortCause?
-                        TransactionalContext.getCurrentContext());
+                try {
+                    if (TransactionalContext.isInTransaction()) {
+                        // If this transaction has entries that wish to capture the committed address
+                        // invoke its preCommitCallbacks with the tokenResponse from the sequencer.
+                        // Note that we might invoke the same method multiple times on retries,
+                        // which means the preCommitCallback must be idempotent.
+                        TokenResponse finalTokenResponse = tokenResponse;
+                        log.debug("append: Invoking {} preCommitListeners",
+                                TransactionalContext.getRootContext().getPreCommitListeners().size());
+                        TransactionalContext.getRootContext()
+                                .getPreCommitListeners()
+                                .forEach(e -> e.preCommitCallback(finalTokenResponse));
+                    }
+
```

Comment:
This doesn't look correct. 

Since the SerializationHandle is acquired before the `for-loop` retries that will end up calling `preCommitCallback` will change the payload, but it will not be re-serialized. Only the token portion will be. This can introduce correctness issues to the queue ordering. 

RIP. I don't remember seeing a PR that introduced this code. It seems like it went in #2498, but it has no reviews.

Reply:
#2498 is the backporting, the PR on master went through reviews.
A quick fix can be putting the serialization in for-loop, but I think it's not the best way of doing it, because we really don't need to serialize payload on every retry just to make it work for the queue case.

@Maithem @hisundar Let's have a discussion to see what is a better solution.

Reply:
I see. 

Even pre-this patch this call back can introduce a correctness issue. Imagine getting a commit token for a transaction, but then before replicating the call back changes the payload's contents such that the initial transaction resolution step is invalid.